File: routes/calls.js
import { Router } from 'express';
import { seniorService } from '../services/seniors.js';
import { requireAuth, requireAdmin } from '../middleware/auth.js';
import { callLimiter } from '../middleware/rate-limit.js';
import { idempotencyMiddleware } from '../middleware/idempotency.js';
import { validateBody } from '../middleware/validate.js';
import { initiateCallSchema } from '../validators/schemas.js';
import { canAccessSenior, routeError } from './helpers.js';
import { logAudit, authToRole } from '../services/audit.js';
import { sendError } from '../lib/http-response.js';
import { endTelnyxCall, initiateTelnyxOutboundCall } from '../services/telnyx.js';

const router = Router();

function formatPhoneForCall(phone) {
  const digits = String(phone || '').replace(/\D/g, '');
  if (!digits) return null;
  if (digits.length === 10) return `+1${digits}`;
  if (digits.length === 11 && digits.startsWith('1')) return `+${digits}`;
  return String(phone || '').startsWith('+') ? String(phone) : `+${digits}`;
}

// API: Initiate outbound call (strict rate limit: 5/min)
router.post('/api/call', requireAuth, validateBody(initiateCallSchema), idempotencyMiddleware, callLimiter, async (req, res) => {
  const { seniorId } = req.body;
  const PIPECAT_URL = req.app.get('baseUrl');

  logAudit({
    userId: req.auth.userId,
    userRole: authToRole(req.auth),
    action: 'create',
    resourceType: 'call',
    ipAddress: req.ip,
    userAgent: req.get('user-agent'),
    metadata: { seniorId },
  });

  try {
    const senior = await seniorService.getById(seniorId);
    if (!senior) {
      return sendError(res, 404, { error: 'Senior not found' });
    }
    if (!senior.isActive) {
      return sendError(res, 400, { error: 'Senior is not active' });
    }

    // Check if user can access this senior before resolving/calling the phone number.
    if (!await canAccessSenior(req.auth, senior.id)) {
      return sendError(res, 403, { error: 'Access denied to this senior' });
    }

    const callPhone = formatPhoneForCall(senior.phone);
    if (!callPhone) {
      return sendError(res, 400, { error: 'Senior phone is not callable' });
    }

    const call = await initiateTelnyxOutboundCall({
      seniorId: senior.id,
      callType: 'check-in',
      baseUrl: PIPECAT_URL,
    });
    res.json({
      success: true,
      provider: 'telnyx',
      callSid: call.callSid,
      callControlId: call.callControlId,
      seniorId: senior.id,
    });

  } catch (error) {
    routeError(res, error, 'POST /api/call');
  }
});

// API: List active calls (admin only)
// Active calls are tracked by Pipecat — query its /health endpoint
router.get('/api/calls', requireAdmin, (req, res) => {
  res.json({
    activeCalls: 0,
    callSids: [],
  });
});

// API: End a call (admin only)
router.post('/api/calls/:callSid/end', requireAdmin, async (req, res) => {
  try {
    await endTelnyxCall(req.params.callSid, { baseUrl: req.app.get('baseUrl') });
    res.json({ success: true, provider: 'telnyx' });
  } catch (error) {
    routeError(res, error, 'POST /api/calls/:callSid/end');
  }
});

export default router;

File: pipecat/api/routes/calls.py
"""Call management API routes for the Telnyx voice path."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Request
from loguru import logger

from api.middleware.auth import require_auth, require_admin, AuthContext
from api.routes.call_context import call_metadata
from api.routes.telnyx import TelnyxOutboundCallRequest, create_telnyx_outbound_call, end_telnyx_call
from api.validators.schemas import InitiateCallRequest
from services.audit import fire_and_forget_audit, auth_to_role

router = APIRouter()


def _format_phone_for_call(phone: str) -> str:
    digits = "".join(ch for ch in str(phone or "") if ch.isdigit())
    if not digits:
        return ""
    if len(digits) == 10:
        return f"+1{digits}"
    if len(digits) == 11 and digits.startswith("1"):
        return f"+{digits}"
    if str(phone or "").startswith("+"):
        return str(phone)
    return f"+{digits}"


@router.post("/api/call")
async def initiate_call(
    request: Request,
    body: InitiateCallRequest,
    auth: AuthContext = Depends(require_auth),
):
    """Initiate an outbound call for an authorized senior.

    Pipecat creates the Telnyx call and hydrates context on the voice service.
    """
    senior_id = body.senior_id

    fire_and_forget_audit(
        user_id=auth.user_id,
        user_role=auth_to_role(auth),
        action="create",
        resource_type="call",
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
        metadata={"senior_id": senior_id},
    )
    try:
        from services.seniors import get_by_id

        senior = await get_by_id(senior_id)
        if not senior:
            raise HTTPException(status_code=404, detail="Senior not found")
        if not senior.get("is_active", True):
            raise HTTPException(status_code=400, detail="Senior is not active")

        # Access control for non-admin users
        if not auth.is_admin:
            if not auth.clerk_user_id:
                raise HTTPException(status_code=403, detail="Access denied to this senior")
            from services.caregivers import can_access_senior
            has_access = await can_access_senior(auth.clerk_user_id, senior["id"])
            if not has_access:
                raise HTTPException(status_code=403, detail="Access denied to this senior")

        phone = _format_phone_for_call(senior["phone"])
        if not phone:
            raise HTTPException(status_code=400, detail="Senior phone is not callable")

        call = await create_telnyx_outbound_call(
            TelnyxOutboundCallRequest(seniorId=senior_id, callType="check-in")
        )

        logger.info("Initiated Telnyx call {sid} for senior {senior_id}", sid=call["callSid"], senior_id=senior["id"])
        return {
            "success": True,
            "provider": "telnyx",
            "call_sid": call["callSid"],
            "call_control_id": call["callControlId"],
            "senior_id": senior["id"],
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error("Failed to initiate call: {err}", err=str(e))
        raise HTTPException(status_code=500, detail="Failed to initiate call")


@router.get("/api/calls")
async def list_active_calls(request: Request, auth: AuthContext = Depends(require_admin)):
    """List active calls (admin only)."""
    fire_and_forget_audit(
        user_id=auth.user_id,
        user_role=auth_to_role(auth),
        action="read",
        resource_type="call",
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
    )
    return {
        "active_calls": len(call_metadata),
        "call_sids": list(call_metadata.keys()),
    }


@router.post("/api/calls/{call_sid}/end")
async def end_call(call_sid: str, auth: AuthContext = Depends(require_admin)):
    """End an active call (admin only)."""
    try:
        return await end_telnyx_call(call_sid)
    except Exception as e:
        logger.error("Failed to end call {sid}: {err}", sid=call_sid, err=str(e))
        raise HTTPException(status_code=500, detail="Failed to end call")
